localstack-core/localstack/services/s3/provider.py

```diff
@@ -700,6 +708,15 @@ def put_object(
             body = AwsChunkedDecoder(body, decoded_content_length, s3_object=s3_object)
 
         with self._storage_backend.open(bucket_name, s3_object, mode="w") as s3_stored_object:
+            # as we are inside the lock here, if multiple concurrent requests happen for the same object, it's the first
+            # one to finish to succeed, and subsequent will raise exceptions. Once the first write finishes, we're
+            # opening the lock and other requests can check this condition
+            if if_none_match and object_exists_for_precondition_write(s3_bucket, key):
+                raise PreconditionFailed(
+                    "At least one of the pre-conditions you specified did not hold",
+                    Condition="If-None-Match",
+                )
```

Comment:
It's a good question 😄 the `open` call is also linked with some read/write locking mechanism, so if 2 concurrent operations were to call `PutObject`, as we call `open()` in `w` mode, the lock is allowing only one request to write. (In read, as many as we want can open it, as long as there is no write underway). 

Meaning that if we do the call in this block, it allows us to enjoy this locking mechanism and make sure that if there were 2 concurrent writes, we won't get into this block before the write is over.

If we were to set the check before the block, it could go this way:
![Untitled-2024-08-22-0215](https://github.com/user-attachments/assets/795df641-39a9-4e33-af4d-a2b632e56ec5)

If the check is in before, both will say "ok", but then `Request 2` is going to overwrite the file. If we do it in the lock, the file is created by `Request 1` and `Request 2` will see that there is an existing file, and fail and not overwrite. 

Hopefully this is clear?

Reply:
Awesome! thanks for the clarification, it makes a lot of sense! Good work on the s3 provider code!